src/main/java/io/reactivex/internal/schedulers/ScheduledRunnable.java

```diff
@@ -66,13 +71,13 @@ public void run() {
         } finally {
             lazySet(THREAD_INDEX, null);
             Object o = get(PARENT_INDEX);
-            if (o != DISPOSED && o != null && compareAndSet(PARENT_INDEX, o, DONE)) {
+            if (o != PARENT_DISPOSED && compareAndSet(PARENT_INDEX, o, DONE) && o != null) {
```

Comment:
Change of `compareAndSet` position in `if` statement seems to be a behavior change…

Though it doesn't affect `ScheduledRunnableTest` on my machine

Do we need that change?

Reply:
Yes. With the previous order, if `o` was `null` because of lack of a parent, the CAS didn't happen and the task would partially appear to be still active. With the sync-async marker changes, `isDisposed` was switched to check the parent reference for indication of being disposed instead of the future reference, which required now 3 comparisons instead of 2.

Reply:
Makes sense, can you please add a test to cover that then? 

When I changed order back to original one on your branch, whole `ScheduledRunnableTest` passed

Reply:
Sure.

Reply:
Unit tests added.